### PR TITLE
Refine vacations pipeline (Part 1)

### PR DESCRIPTION
**Pull request by sponkrashin**
Issue #291

### CODE REVIEW

File: server/Arcadia.Assistant.Calendar/Vacations/EmployeeVacationsActor.cs [hunk 1]
```diff
@@ -59,9 +82,58 @@ protected override void OnCommand(object message)
...
                     this.vacationsRegistry
-                        .Ask<VacationsRegistry.GetVacationInfo.Response>(new VacationsRegistry.GetVacationInfo(this.EmployeeId))
+                        .Ask<VacationsRegistry.GetVacationInfo.Response>(
+                            new VacationsRegistry.GetVacationInfo(this.EmployeeId),
+                            this.timeoutSetting)
                         .ContinueWith(x => new GetVacationsCredit.Response(x.Result.VacationsCredit))
                         .PipeTo(this.Sender);
+                    break;
+
+                case VacationApproveMessage msg:
```
Comment (santee) [inline server/Arcadia.Assistant.Calendar/Vacations/EmployeeVacationsActor.cs]: -> ApproveVacation

File: server/Arcadia.Assistant.Calendar/Vacations/EmployeeVacationsActor.cs [hunk 2]
```diff
@@ -59,9 +82,58 @@ protected override void OnCommand(object message)
...
+                    var @event = new VacationIsApprovedOnce
```
Comment (santee) [inline server/Arcadia.Assistant.Calendar/Vacations/EmployeeVacationsActor.cs]: methods please :)

File: server/Arcadia.Assistant.Calendar/Vacations/Events/VacationIsApprovedOnce.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿namespace Arcadia.Assistant.Calendar.Vacations.Events
+{
+    using System;
+    using System.Runtime.Serialization;
+
+    [DataContract]
+    public class VacationIsApprovedOnce
```
Comment (santee) [inline server/Arcadia.Assistant.Calendar/Vacations/Events/VacationIsApprovedOnce.cs]: -> UserGrantedVacationApproval ?

File: server/Arcadia.Assistant.Calendar/Vacations/Events/VacationIsApprovedOnce.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        [DataMember]
+        public string EventId { get; set; }
+
+        [DataMember]
+        public DateTimeOffset TimeStamp { get; set; }
+
+        [DataMember]
+        public string UserId { get; set; }
```
Comment (santee) [inline server/Arcadia.Assistant.Calendar/Vacations/Events/VacationIsApprovedOnce.cs]: I don't think we put UserId anywhere. It is employee id, right? So it should be ApproverId

File: server/Arcadia.Assistant.Organization.Abstractions/VacationApprovalsChecker.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+            return Context.DI().Props<VacationApprovalsChecker>();
+        }
+
+        protected override void OnReceive(object message)
+        {
+            switch (message)
+            {
+                case GetNextVacationRequestApprover msg:
+                    this.GetNextApprover(msg.EmployeeId, msg.ExistingApprovals)
+                        .PipeTo(this.Sender, success: r => new GetNextVacationRequestApprover.Response(r));
```
Comment (santee) [inline server/Arcadia.Assistant.Organization.Abstractions/VacationApprovalsChecker.cs]: add a Faliure? So it won't fail with a timeout when database fails with an error

Comment (santee) [reply server/Arcadia.Assistant.Organization.Abstractions/VacationApprovalsChecker.cs]: I think now it sends a generic Failure message back to sender, which is not handled, and your code waits for Response than never happens and things go south :)

File: server/Arcadia.Assistant.Organization/OrganizationActor.cs [hunk 6]
```diff
@@ -52,6 +51,10 @@ protected override void OnReceive(object message)
                     this.employeesActor.Forward(query);
                     break;
 
+                case GetNextVacationRequestApprover msg:
```
Comment (santee) [inline server/Arcadia.Assistant.Organization/OrganizationActor.cs]: in what scenario is that used?

Comment (sponkrashin) [reply server/Arcadia.Assistant.Organization/OrganizationActor.cs]: https://github.com/SoftwareCountry/arcadian-assistant/pull/299/files/077466f3825bf7a890237136345194404833c50b#diff-89e87f12de8409707af9e31c732e4cedR110

Comment (sponkrashin) [reply server/Arcadia.Assistant.Organization/OrganizationActor.cs]: Legacy :)

File: server/Arcadia.Assistant.Calendar/Vacations/EmployeeVacationsActor.cs (hunk 2 above)
Comment (sponkrashin) [reply server/Arcadia.Assistant.Calendar/Vacations/EmployeeVacationsActor.cs]: Done.

File: server/Arcadia.Assistant.Calendar/Vacations/Events/VacationIsApprovedOnce.cs (hunk 3 above)
Comment (sponkrashin) [reply server/Arcadia.Assistant.Calendar/Vacations/Events/VacationIsApprovedOnce.cs]: Done.

File: server/Arcadia.Assistant.Calendar/Vacations/Events/VacationIsApprovedOnce.cs (hunk 4 above)
Comment (sponkrashin) [reply server/Arcadia.Assistant.Calendar/Vacations/Events/VacationIsApprovedOnce.cs]: Done.

File: server/Arcadia.Assistant.Organization/OrganizationActor.cs (hunk 6 above)
Comment (sponkrashin) [reply server/Arcadia.Assistant.Organization/OrganizationActor.cs]: Removed.

File: server/Arcadia.Assistant.CSP/Vacations/CspVacationApprovalsChecker.cs [hunk 7]
```diff
@@ -37,16 +37,17 @@ protected override async Task<string> GetNextApprover(string employeeId, IEnumer
             var parentDepartment = allDepartments.FirstOrDefault(d => d.DepartmentId == employeeDepartment.ParentDepartmentId);
             var isEmployeeChief = employeeDepartment.ChiefId == employeeId;
 
-            var neededApprovals = new List<string>();
+            string mainApproverId = null;
+            string secondApproverId = null;
```
Comment (santee) [inline server/Arcadia.Assistant.CSP/Vacations/CspVacationApprovalsChecker.cs]: so there won't be three approvers? :) what you can have is finalApprover / preapprovers

File: server/Arcadia.Assistant.Organization.Abstractions/VacationApprovalsChecker.cs (hunk 5 above)
Comment (santee) [reply server/Arcadia.Assistant.Organization.Abstractions/VacationApprovalsChecker.cs]: Like you tried with Response, Success: Response, Failure: Response before.

**Comment by sponkrashin** [comment]
I made it as simple as possible, having in mind that it is temporary
solution anyway.

чт, 29 нояб. 2018 г., 16:32 Alexander Shevnin notifications@github.com:

File: server/Arcadia.Assistant.Organization.Abstractions/VacationApprovalsChecker.cs (hunk 5 above)
Comment (santee) [reply server/Arcadia.Assistant.Organization.Abstractions/VacationApprovalsChecker.cs]: Then you can handle it. I think now you just await it via Ask? So it'll fail with timeout? Here, you'll be able to handle it properly

File: server/Arcadia.Assistant.CSP/Vacations/CspVacationApprovalsChecker.cs (hunk 7 above)
Comment (sponkrashin) [reply server/Arcadia.Assistant.CSP/Vacations/CspVacationApprovalsChecker.cs]: I made it as simple as possible, having in mind that it is temporary
solution anyway.

File: server/Arcadia.Assistant.Organization.Abstractions/VacationApprovalsChecker.cs (hunk 5 above)
Comment (sponkrashin) [reply server/Arcadia.Assistant.Organization.Abstractions/VacationApprovalsChecker.cs]: Yes, I understand. But what's next? How application will know about next approver, if this request returned an error?

**Pull request merged**